theano/tensor/nnet/abstract_conv.py

```diff
@@ -177,6 +179,383 @@ def conv2d(input,
+        to as a convolution, and this is the default. If ``False``, the filters
+        are not flipped and the operation is referred to as a
+        cross-correlation.
+
+    :rtype: symbolic 4D tensor.
+    :return: set of feature maps generated by convolutional layer. Tensor is
+        of shape (batch size, output channels, output rows, output columns)
+
+    :note: If CuDNN is available, it will be used on the
+        GPU. Otherwise, it is the *CorrMM* convolution that will be used
+        "caffe style convolution".
+
+    :note: This is only supported in Theano 0.8 or the development
+        version until it is released.
+
+    """
+    gradWeight_op = AbstractConv2d_gradWeights(imshp=input_shape,
+                                               kshp=filter_shape,
+                                               border_mode=border_mode,
+                                               subsample=subsample,
+                                               filter_flip=filter_flip)
+
+    return gradWeight_op(input, output_grad, input_shape)
+
+
+def bilinear_kernel_2D(ratio, normalize=True):
+    """ This function gets a ratio
+    and returns the normalized 2D filters that can
+    be applied to any image to upsample it by the ratio
+    using bilinear interpolation.
+
+    :type ratio: integer
+    : param ratio: the ratio by which an image is upsampled
+        in the 2D space.
+
+    :type normalize: bool
+    : param normalize: indicates whether to normalize the
+    kernel or not. Default is True.
+
+    returns: the 2D kernels that can be applied to
+             any given image to upsample it by indicated ratio
+             using bilinear interpolation.
+    """
+    assert ratio > 1
+    kern_size = ratio * 2 - 1
+    kern = np.zeros((kern_size, kern_size))
+    for ii in range(-ratio + 1, ratio):
+        for jj in range(-ratio + 1, ratio):
+            kern[ii + ratio - 1, jj + ratio - 1] = (ratio - abs(ii)) * (ratio - abs(jj))
+    if normalize:
+        kern /= ratio**2
+    return kern
+
+
+def bilinear_kernel_1D(ratio, normalize=True):
+    """ This function gets a ratio
+    and returns the normalized 1D filters that can
+    be applied to any image to upsample it by the ratio
+    using bilinear interpolation.
+
+    :type ratio: integer
+    : param ratio: the ratio by which an image is upsampled
+        in the 1D space.
+
+    :type normalize: bool
+    : param normalize: indicates whether to normalize the
+    kernel or not. Default is True.
+
+    returns: the 1D kernels that can be applied to
+             any given image to upsample it by indicated ratio
+             using bilinear interpolation in one dimension.
+    """
+    assert ratio > 1
+    kern_size = ratio * 2 - 1
+    kern = np.zeros((kern_size))
+    for ii in range(-ratio + 1, ratio):
+        kern[ii + ratio - 1] = ratio - abs(ii)
+    if normalize:
+        kern /= ratio
+    return kern
+
+
+def bilinear_upsampling(input,
```

Comment:
Please use a different name, as this will redefine the standard python function [`input`](https://docs.python.org/3/library/functions.html#input).